Bugfix/2260 fail to render crds

alexandre-allard opened the pull request:
**Component**: salt, solutions

**Context**: 

**Summary**:

**Acceptance criteria**: 


---

Closes: #2260

bert-e commented:
# Hello alexandre-allard-scality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/2260-fail-to-render-crds` into
`development/2.4`
* [w/2.5/bugfix/2260-fail-to-render-crds](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/bugfix/2260-fail-to-render-crds) will be merged into `development/2.5` 
* [w/2.6/bugfix/2260-fail-to-render-crds](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/bugfix/2260-fail-to-render-crds) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

slaperche-scality commented on salt/_modules/metalk8s.py:
```diff
@@ -275,3 +275,14 @@ def check_pillar_keys(keys, refresh=True, pillar=None, raise_error=True):
+    """
+
+    with open(path, 'r') as fd:
+        return fd.read()
```
> passing the result to the yaml filter is failing, because it sees the content as a single string

That's weird, because here you return a single string as well so why does it work better?

alexandre-allard replied:
Well, I must admit that the commit message is not really good.
But, Salt is returning something that's not parsable by the yaml filter.
It's like the whole content is escaped with quotes, the YAML filter sees only a single string, instead of a complex object and renders it badly.
Anyway, I've removed this part. :)

gdemonet reviewed (commented):
I'm not sure that's the problem.

It's not a problem of Jinja markup, but YAML markup. We were trying to espace the YAML data as a string, using:

```jinja
module.run:
  - state.template_str:
    - tem: "{{ sls_content | yaml }}"
```

If the YAML file contains quotes `"`, then the escaping will stop mid-string and the resulting YAML will be invalid.

IMO, inlining the file, which comes from an unknown source, isn't a great idea.
From what I understand, we still want to gather the file from the bootstrap minion, because we may not have the files in the Salt Master container. Honestly, I think having access to `/srv/scality` entirely in the Salt Master container would be a much more acceptabe solution.

If this is true, then you could replace the whole `file_read` with a simple:

```jinja
  {%- for crd_file in crd_files %}
{{ action }} CRD "{{ crd_file }}" for Solution {{ solution.name }}:
  metalk8s_kubernetes.object_present:
    - name: {{ crd_file }}

  {%- endfor %}
```

slaperche-scality commented on salt/metalk8s/orchestrate/solutions/deploy-components.sls:
```diff
@@ -37,16 +37,16 @@ include:
-      - tem: "{{ sls_content | yaml }}"
+  {{ state }}:
+    - manifest:
+        {{ sls_content | yaml(False) | indent(8) }}"
```
What is the use of `False` here?

alexandre-allard replied:
It corresponds to the `default_flow_style=False` argument of the `yaml.dump()` function in Python.

alexandre-allard commented:
Ok, it wasn't really clear in the issue, I though it was due to some Jinja markup as written in my commit message.
My first idea was to use the `name` argument but the file is not accessible on the master, I'll try to mount the whole `/srv/scality` directory in the salt master container.

alexandre-allard committed:
salt: mount solution directory in salt master

this change will ease the management and deployment
of solutions as the salt-master is able to read
the resources from the solutions archives

Refs: #2260

alexandre-allard committed:
salt: use metalk8s kubernetes state instead of renderer

We use the metalk8s_kubernetes state instead of the
kubernetes renderer, otherwise it may fails while
parsing the CRDs yaml.
It also makes what is done clearer and simpler
than before.

Refs: #2260

bert-e commented:
# History mismatch



Merge commit #f310a656bbedfcf6aada317011b6c981f53fa25d on the integration branch
`w/2.5/bugfix/2260-fail-to-render-crds` is merging a branch which is neither the current
branch `bugfix/2260-fail-to-render-crds` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `bugfix/2260-fail-to-render-crds` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

alexandre-allard commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/2260-fail-to-render-crds` into
`development/2.4`
* [w/2.5/bugfix/2260-fail-to-render-crds](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/bugfix/2260-fail-to-render-crds) will be merged into `development/2.5` 
* [w/2.6/bugfix/2260-fail-to-render-crds](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/bugfix/2260-fail-to-render-crds) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

gdemonet reviewed (dismissed):
LGTM, what do you think @NicolasT @TeddyAndrieux?

bert-e commented:
# History mismatch



Merge commit #704776acd412d0de2e7fbc24d65f26312c8b31d7 on the integration branch
`w/2.5/bugfix/2260-fail-to-render-crds` is merging a branch which is neither the current
branch `bugfix/2260-fail-to-render-crds` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `bugfix/2260-fail-to-render-crds` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

alexandre-allard committed:
scripts: reconfigure salt master when importing Solution

The Solution archives are now mounted inside the
Salt master container to ease the access to the
solutions resource definitions (e.g. CRDs)

Refs: #2260

alexandre-allard commented:
We could also mount each and every solution archives at import/activation time, but it means restarting the salt master container every time there is a new solution or when we remove an old one.

alexandre-allard commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/2260-fail-to-render-crds` into
`development/2.4`
* [w/2.5/bugfix/2260-fail-to-render-crds](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/bugfix/2260-fail-to-render-crds) will be merged into `development/2.5` 
* [w/2.6/bugfix/2260-fail-to-render-crds](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/bugfix/2260-fail-to-render-crds) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

bert-e commented:
# History mismatch



Merge commit #505a59a04ecaa071fc4b52a2c6facf1cbbf39a4f on the integration branch
`w/2.5/bugfix/2260-fail-to-render-crds` is merging a branch which is neither the current
branch `bugfix/2260-fail-to-render-crds` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `bugfix/2260-fail-to-render-crds` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

alexandre-allard commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/2260-fail-to-render-crds` into
`development/2.4`
* [w/2.5/bugfix/2260-fail-to-render-crds](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/bugfix/2260-fail-to-render-crds) will be merged into `development/2.5` 
* [w/2.6/bugfix/2260-fail-to-render-crds](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/bugfix/2260-fail-to-render-crds) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

bert-e commented:
# History mismatch



Merge commit #fcb280a88ec3c60268103b0155f4ae5f4881802e on the integration branch
`w/2.5/bugfix/2260-fail-to-render-crds` is merging a branch which is neither the current
branch `bugfix/2260-fail-to-render-crds` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `bugfix/2260-fail-to-render-crds` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

alexandre-allard commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/2260-fail-to-render-crds` into
`development/2.4`
* [w/2.5/bugfix/2260-fail-to-render-crds](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/bugfix/2260-fail-to-render-crds) will be merged into `development/2.5` 
* [w/2.6/bugfix/2260-fail-to-render-crds](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/bugfix/2260-fail-to-render-crds) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort

alexandre-allard commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @tcarmet

* @jbertran

* @carlito-scality

* @slaperche-scality

* @nicolast

* @teddyandrieux

* @monpote

* @sayf-eddine-scality

* @nootal

* @alexandre-allard-scality

* @gdemonet

* @lucieleonard

* @thomasdanan

* @cmonfort










*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye alexandre-allard-scality.
